modules/afsql/afsql.c

```diff
@@ -862,53 +829,49 @@ afsql_dd_build_insert_command(AFSqlDestDriver *self, LogMessage *msg, GString *t
 static inline gboolean
 afsql_dd_is_transaction_handling_enabled(const AFSqlDestDriver *self)
 {
-  return self->flush_lines_queued != -1;
+  return !!(self->flags & AFSQL_DDF_EXPLICIT_COMMITS);
 }
 
 static inline gboolean
 afsql_dd_should_begin_new_transaction(const AFSqlDestDriver *self)
 {
-  return self->flush_lines_queued == 0;
+  return afsql_dd_is_transaction_handling_enabled(self) && self->super.batch_size == 1;
 }
 
 static inline gboolean
 afsql_dd_should_commit_transaction(const AFSqlDestDriver *self)
 {
-  return afsql_dd_is_transaction_handling_enabled(self) && self->flush_lines_queued == self->flush_lines;
+  return afsql_dd_is_transaction_handling_enabled(self) && self->super.batch_size >= self->flush_lines;
 }
 
-static inline void
-afsql_dd_rollback_msg(AFSqlDestDriver *self, LogMessage *msg, LogPathOptions *path_options)
-{
-  if (self->flags & AFSQL_DDF_EXPLICIT_COMMITS)
-    log_queue_rewind_backlog(self->queue, 1);
-  else
-    log_queue_push_head(self->queue, msg, path_options);
-}
-
-static inline gboolean
-afsql_dd_handle_insert_row_error_depending_on_connection_availability(AFSqlDestDriver *self,
-    LogMessage *msg,
-    LogPathOptions *path_options)
+static worker_insert_result_t
+afsql_dd_handle_insert_row_error_depending_on_connection_availability(AFSqlDestDriver *self)
 {
   const gchar *dbi_error, *error_message;
 
+#if 0
+
+  /* FIXME */
   if (dbi_conn_ping(self->dbi_ctx) == 1)
     {
-      afsql_dd_rollback_msg(self, msg, path_options);
-      return TRUE;
+      /* FIXME: I am not sure it makes sense for this behavior, the SQL
+       * server returned an error for our SQL query and the error is not
+       * about a dropped SQL connection, so most probably our SQL has syntax
+       * errors or violates some constraints.  Earlier this just dropped and
+       * retried a single message a few times and then progressed.  */
+
+      return WORKER_INSERT_RESULT_SINGLE_ERROR;
```

Comment:
With @bazsi we concluded that it would worth return WORKER_INSERT_RESULT_ERROR here, that would retry the batch 3 times. The original implementation retried only the last message in the batch, but this change is reasonable.